File: lua/problem_79.lua
local M = {}

function M.can_be_non_decreasing(nums)
    local a = {table.unpack(nums)}
    local modified = false
    for i = 2, #a do
        if a[i] < a[i-1] then
            if modified then return false end
            modified = true
            if i >= 3 and a[i] < a[i-2] then a[i] = a[i-1]
            else a[i-1] = a[i] end
        end
    end
    return true
end

function M.problem79()
    print("Problem 79:")
    assert(    M.can_be_non_decreasing({10,5,7}))
    assert(not M.can_be_non_decreasing({10,5,1}))
    assert(    M.can_be_non_decreasing({1,2,3}))
    print("All Tests Passed")
end

return M

File: python/problem_79.py
from typing import List


def can_be_non_decreasing(nums: List[int]) -> bool:
    nums = nums[:]
    modified = False
    for i in range(1, len(nums)):
        if nums[i] < nums[i-1]:
            if modified:
                return False
            modified = True
            if i >= 2 and nums[i] < nums[i-2]:
                nums[i] = nums[i-1]
            else:
                nums[i-1] = nums[i]
    return True


def main():
    print("Problem 79:")
    assert     can_be_non_decreasing([10, 5, 7])
    assert not can_be_non_decreasing([10, 5, 1])
    assert     can_be_non_decreasing([1, 2, 3])
    print("All Tests Passed")
